### DISCUSSION TITLE
How can I config my pulsar to send large message more efficiency?

### DISCUSSION THREAD

**Discussion by wenze159632** [apache/pulsar#21700] [Q&A] [answered]
when I try to use OpenMessaging Benchmark to test my Pulsar, I find it performance is unacceptable.
The performance is excellent when configured with a 1kb size and generating 2400 records per second, but experiences severe performance degradation when configured with a 24kb size and generating 100 records per second.
here is the result of 1kb, 2400msg/s
![1702128794673](https://github.com/apache/pulsar/assets/153369683/46d2e9b5-8e9a-48de-8b24-67ac1440c400)

and then 24kb,100msg/s
![1702128743059](https://github.com/apache/pulsar/assets/153369683/f6d9c0d2-d77d-412d-aa87-14fe74c82e49)

My hard drive is an SSD with a read/write speed of up to 100MB/s, and you have 64GB of memory. The operating system  is Ubuntu 22.0. 
this is my OMB config 
```bash
name: 1 topic / 1 partition / 100b

topics: 1
partitionsPerTopic: 1
messageSize: 1024
payloadFile: "payload/payload-1Kb.data"

subscriptionsPerTopic: 1
consumerPerSubscription: 1
producersPerTopic: 1
producerRate: 2400
consumerBacklogSizeGB: 0
testDurationMinutes: 5
```
my pulsar use default configuration

**Comment by merlimat** [comment, marked as answer]
I would start by checking that the BookKeeper data is actually written in the intended disks. 

 Are you separating the BK journal and ledger device to different disks? Can you run `iostat -xm 10` in the BookKeeper node to verify the iops on the disks

**Reply by wenze159632** [reply]
thinks for your advise，but I don't think it's a hard drive read/write issue. I just set the messages sent by OMB not to be written to the disk (i.e., modify the file below), but the problem still exists. 
![1702188326623](https://github.com/apache/pulsar/assets/153369683/92c403a0-535d-44a7-b7e9-20d36acce600)
In this case ,I use the `iotop`  command and  find that the IO resource usage on the bookie server is below 1%.

In addition, I also found that his test results fluctuate dramatically, with test results varying by several thousand times.
![1702189008944](https://github.com/apache/pulsar/assets/153369683/bcb5bcd8-cb7f-4288-b2ab-89b18e423138)

**Reply by merlimat** [reply]
The change you made in the driver code doesn't change the fact that data is persisted on disk

**Reply by merlimat** [reply]
If you want to test without the disk, you should switch the topic type to "non-persistent" in the driver configuration file

**Reply by wenze159632** [reply]
Your advice was very effective. After I tried changing the message type, the performance of Pulsar improved significantly. （Pub Latency  can be close to 1ms）So, it can be basically confirmed that it's an issue with Bookeeper. This is the result I got from running` iostat -xm 10` as per your suggestion.
![1702265975507](https://github.com/apache/pulsar/assets/153369683/7e1ee605-fb2b-47bf-9154-f8b617df0b0c)
Here's the specific configuration info for my bookeeper.conf
```bash
journalDirectory=/home/leizhou/data/bookkeeper/journal
ledgerDirectories=/home/leizhou/data/bookkeeper/ledgers
```
and here is my hardware config 
![1702266573723](https://github.com/apache/pulsar/assets/153369683/92d43f60-5dd4-475a-9151-4e5486a14fea)
As you can see，I only have 1 SSD hard drive, based on this situation I put ledges and journal in the same folder, do you have any other suggestions. 
Very thanks for your help

**Reply by wenze159632** [reply]
I just redeployed Pulsar on a new machine, and I distributed the journal and ledgers on two different disks as you suggested, but it didn't work very well, do you have any other suggestions?
1000msg/s，24kb in pulsar-perf,produce
![1702300750168](https://github.com/apache/pulsar/assets/153369683/2034a434-99bc-42b9-861c-cb4e49d31b8d)
and this  the Specific data of it
![1702300972628](https://github.com/apache/pulsar/assets/153369683/e9772407-2262-4377-8dcf-91ae9aed1609)

besides， here is pulsar-perf,cosume
![1702301174513](https://github.com/apache/pulsar/assets/153369683/daf47da0-8ac8-47ad-82ff-636febe95afd)

**Comment by wenze159632** [comment]
I configured the new NVIDIMM hard drive and used it with the original SSD hard drive, and finally now the test result of 24Kb, 1000 messages has reached the 99.99th percentile of about 7ms, thanks for the answer

**Reply by wenze159632** [reply]
Another important thing to note is that a Zookeeper cluster should be deployed on servers with high I/O performance, as it greatly enhances the performance of the cluster.
